File: sdk/tools/bakery/src/bakery/py_runtime.py
"""Python WASM runtime install/check helpers.

Pie's Python inferlets are built with `factored-componentize-py` — a
"factored" build that imports a `componentize-py-runtime` core module
from the host instead of bundling the whole interpreter into each WASM.
The host loads that core module from `$PIE_HOME/py-runtime/shared/*.wasm`,
so the runtime tarball must be present before any Python inferlet can
actually run.

This module is shared between bakery (which fetches at build time so
the WASM is immediately runnable) and the pie-server engine (which
fetches lazily on startup). Both call paths converge on
`ensure_installed()`.
"""

from __future__ import annotations

import lzma
import os
import tarfile
from io import BytesIO
from pathlib import Path

import httpx

# Pinned to the version pie-server expects. Bumping this is a coordinated
# change with the runtime/program/python module on the Rust side.
RUNTIME_URL = (
    "https://registry.pie-project.org/api/v1/runtimes/python3.14/0.3.0/download"
)


def get_pie_home() -> Path:
    """Resolve $PIE_HOME, falling back to ~/.pie."""
    if pie_home := os.environ.get("PIE_HOME"):
        return Path(pie_home)
    return Path.home() / ".pie"


def get_runtime_dir() -> Path:
    """The directory the host expects to find shared/*.wasm in."""
    return get_pie_home() / "py-runtime"


def is_installed() -> bool:
    """True if a usable py-runtime tree exists.

    We check for `shared/componentize-py-runtime.wasm` specifically — the
    one core module the host *must* link against. If that file is present
    the rest of the tree is almost certainly intact.
    """
    return (get_runtime_dir() / "shared" / "componentize-py-runtime.wasm").is_file()


def ensure_installed(*, quiet: bool = False) -> Path:
    """Install the Python WASM runtime if it isn't already.

    Returns the resolved runtime directory. Raises on download/extract
    failure — callers decide whether to surface the failure immediately
    or continue (e.g. bakery only needs it at run time).

    Args:
        quiet: If True, suppress the progress bar and "extracting" line.
            Used by the engine startup path where we don't want to clutter
            the terminal unless something interesting happens.
    """
    runtime_dir = get_runtime_dir()
    if is_installed():
        return runtime_dir

    pie_home = get_pie_home()
    pie_home.mkdir(parents=True, exist_ok=True)

    blob = _fetch(quiet=quiet)
    decompressed = lzma.decompress(blob)
    with tarfile.open(fileobj=BytesIO(decompressed), mode="r:") as tar:
        tar.extractall(path=pie_home)

    if not is_installed():
        raise RuntimeError(
            f"Python runtime download completed but {runtime_dir} is incomplete."
        )
    return runtime_dir


def _fetch(*, quiet: bool) -> bytes:
    """Download the runtime tarball, optionally with a progress bar.

    Rich is imported lazily so the module stays importable in minimal
    environments (e.g. test harnesses) even though both bakery and
    pie-server already depend on it.
    """
    if quiet:
        with httpx.stream("GET", RUNTIME_URL, follow_redirects=True) as resp:
            resp.raise_for_status()
            return b"".join(resp.iter_bytes())

    from rich.console import Console
    from rich.progress import (
        BarColumn,
        DownloadColumn,
        Progress,
        TextColumn,
        TransferSpeedColumn,
    )

    console = Console()
    chunks = bytearray()
    with httpx.stream("GET", RUNTIME_URL, follow_redirects=True) as resp:
        resp.raise_for_status()
        total = int(resp.headers.get("content-length", 0))
        with Progress(
            TextColumn("[bold blue]{task.description}"),
            BarColumn(),
            DownloadColumn(),
            TransferSpeedColumn(),
        ) as progress:
            task = progress.add_task(
                "Downloading Python 3.14 runtime for Python inferlets",
                total=total or None,
            )
            for chunk in resp.iter_bytes():
                chunks.extend(chunk)
                progress.update(task, advance=len(chunk))
    console.print("[dim]Extracting runtime…[/dim]")
    return bytes(chunks)

File: src/local/py_runtime.rs
use std::path::{Path, PathBuf};

use anyhow::{Result, anyhow, bail};

const RUNTIME_URL: &str =
    "https://registry.pie-project.org/api/v1/runtimes/python3.14/0.4.0/download";

pub fn runtime_dir() -> PathBuf {
    bootstrap::paths::pie_home().join("py-runtime")
}

fn sentinel() -> PathBuf {
    runtime_dir()
        .join("shared")
        .join("componentize-py-runtime.wasm")
}

pub fn is_installed() -> bool {
    sentinel().is_file()
}

pub fn ensure_installed(quiet: bool) -> Result<PathBuf> {
    let dir = runtime_dir();
    if is_installed() {
        return Ok(dir);
    }

    let pie_home = bootstrap::paths::pie_home();
    std::fs::create_dir_all(&pie_home).map_err(|e| anyhow!("create {pie_home:?}: {e}"))?;

    if !quiet {
        eprintln!("Downloading Python WASM runtime from {RUNTIME_URL}…");
    }
    let blob = fetch()?;

    if !quiet {
        eprintln!("Extracting runtime to {}…", pie_home.display());
    }
    extract(&blob, &pie_home)?;

    if !is_installed() {
        bail!(
            "Python runtime download completed but {dir:?} is incomplete \
             (missing {})",
            sentinel().display()
        );
    }
    Ok(dir)
}

pub fn ensure_installed_best_effort(enabled: bool) {
    let dir = runtime_dir();
    if !enabled {
        tracing::debug!(
            "python runtime download skipped: `python_runtime = false` under [sandbox]"
        );
        return;
    }
    if is_installed() {
        tracing::debug!("python runtime already installed at {}", dir.display());
        return;
    }
    tracing::info!(
        "fetching the Python WASM runtime from {RUNTIME_URL} into {}",
        dir.display()
    );
    if let Err(e) = ensure_installed(/*quiet=*/ true) {
        tracing::warn!(
            "could not fetch the Python WASM runtime from {RUNTIME_URL}: {e}. \
             Python inferlets will not run until this download succeeds; set \
             `python_runtime = false` under [sandbox] in the config to skip it."
        );
    }
}

fn fetch() -> Result<Vec<u8>> {
    let resp = reqwest::blocking::Client::new()
        .get(RUNTIME_URL)
        .send()
        .map_err(|e| anyhow!("GET {RUNTIME_URL}: {e}"))?
        .error_for_status()
        .map_err(|e| anyhow!("GET {RUNTIME_URL}: {e}"))?;
    resp.bytes()
        .map(|b| b.to_vec())
        .map_err(|e| anyhow!("read response body: {e}"))
}

fn extract(blob: &[u8], dest: &Path) -> Result<()> {
    let mut decoder = xz2::read::XzDecoder::new(blob);
    let mut tar = tar::Archive::new(&mut decoder);
    tar.unpack(dest)
        .map_err(|e| anyhow!("extract tarball into {dest:?}: {e}"))?;
    Ok(())
}
